### ISSUE TITLE
Running Integration tests on Rococo 

### ISSUE THREAD

**Issue by p5150j** [LibertyDSNP/frequency#1458]
As a team we have Integration tests on a "local node" but another tangible value is to run the Integration tests on Rococo testnet to get a real feel for how they will run on a real network. 

In the past, we build the Integration tests with this in mind but haven't kept up with it so we might have some drift and they might be less agnostic to the network type (hard coded local note, etc)

### The goal of this issue is:

- review the Integration tests and make a list of anything that wouldn't work out of the box on Rococo
- add the files that need to be refactored to be agnostic here (or in sub-task issues)
- get the integration tests to run against Rococo network

Ideas to look into:
https://docs.substrate.io/reference/command-line-tools/try-runtime/

**Comment by johnmela** [comment]
* Possible ideas:
        * Export the chain state from Rococo to create a snapshot in time of the chain; build a local instant seal node to run integration tests against the exported chain state
        * Run the tests directly against the running chain
        * Perhaps there is a valid use case for each? 
            * The first option does not require a testnet account with token that may need continual replenishment.  This could be run by engineers during the development process and/or on CI for PR checks
            * The second option is a more realistic test as it will be running against a live chain that is actively producing blocks. This may be more likely to catch race conditions that we wouldn’t normally catch in a sterile local environment that seals after each block. Tests may need modifications for running in this state

**Comment by johnmela** [comment]
I've been working with this repository:
https://github.com/maxsam4/fork-off-substrate

TLDR; I Suspect it's out of date and will not work without some modification.

The long end of it:
It is a little out of date but there is a more recent branch I have been working out of:
https://github.com/maxsam4/fork-off-substrate/tree/dependabot/npm_and_yarn/polkadot/api-10.0.1
The project is supposed to fork the state of a running chain and bootstrap a new chain to run with said state.

It fails on this RPC call:
https://github.com/maxsam4/fork-off-substrate/blob/dependabot/npm_and_yarn/polkadot/api-10.0.1/index.js#L163
with this error:
RpcError: -32601: RPC call is unsafe to be called externally
An error that I can reproduce when calling the RPC against rococo with polkadot.js UI

I would like to ask Parity in the call this week to determine what a modern method of forking a running chain is.

It may be related to https://docs.substrate.io/reference/command-line-tools/node-template/

the `export-state` command provided in https://docs.substrate.io/reference/command-line-tools/node-template/
may help

**Comment by johnmela** [comment]
I would like to ask Parity in the call this week to determine what a modern method of forking a running chain is.

It may be related to https://docs.substrate.io/reference/command-line-tools/node-template/

the export-state command provided in https://docs.substrate.io/reference/command-line-tools/node-template/
may help

**Comment by johnmela** [comment]
I have paused on Option 1 (running integration tests against a local node with Rococo state). Pulling state from one chain to a node running instant sealing is not turn key functionality and likely will require a fair amount of time investment to achieve.

Option 2 seems to be achievable (in pieces) and may provide more value in the short term. I have a PR open that adds a Makefile target to execute integration tests against Frequency on Rococo or a local relay. Some of the tests  pass against Rococo, however some are failing. The failing tests will need some investigation.

### Failing tests
#### MSA
##### createMsa
1) should fail to add public key if origin does not have an MSA (NoKeyExists)
#### Messages
##### addIPFSMessage
1) should fail if MSA is not valid (InvalidMessageSourceAccount)
2) should successfully add a new public key to an existing MSA & disallow duplicate signed payload submission (SignatureAlreadySubmitted)

**Comment by johnmela** [comment]
Potential follow up issues after discussions with @JoeCap08055 and @rlaferla 
- Timeouts
   - Timeouts are currently set globally and particularly high when executing against Rococo at 200,000ms. 
   - Set default timeout based on a base timeout function that returns the 80% most common timeout value per environment. For example, if 80% of tests complete under 100,000ms in Rococo the function returns 100000. The function would then take an additional optional parameter

- Investigate which tests should / should not be executed against Rococo
    - Which pieces of storage that we set up for Genesis when executing integration tests against a Frequency node with instant seal, should we NOT execute against Rococo
    - Will it cause problems to generate large amounts of unnecessary data against rococo? Example: Schemas could be referenced rather than generated
